### DISCUSSION TITLE
Sample Flow for Keycloak 20

### DISCUSSION THREAD

**Discussion by schenardi** [sventorben/keycloak-home-idp-discovery#160] [Q&A] [answered]
Hi Torben,

thanks for this provider. 
I could succesfully install it but I'm struggling with the configuration of the flow.
Do you perhaps have a complete sample flow?`

Scenario:
User enters username (email oder username), keycloak decides if to forward to openId provider (to Azure AD) > show Azure AD SignIn or to use internal users > show password field.

Thanks; Markus

**Comment by sventorben** [comment]
Hello @schenardi!

sure, the flow from readme should work in your scenario.
![Flow from README.md](https://github.com/sventorben/keycloak-home-idp-discovery/raw/v20.0.3/docs/images/flow.JPG)

Is there any issue with that flow? Let me know if it is not working for you. 

Regards
Sven-Torben

**Reply by schenardi** [reply]
for some reasons it's not working as desired 
I tried to build the flow as in the image ((username and password as sub-flow??) and bound it to my client.
<img width="1285" alt="image" src="https://user-images.githubusercontent.com/22298646/222677811-5a746dfe-93ad-4054-b931-11b41276443f.png">

But the form stays the same > (expected to see one line only with the username?)
<img width="397" alt="image" src="https://user-images.githubusercontent.com/22298646/222677903-012cfb81-6969-49e6-ae20-97d0e73eba3d.png">

Configuration on the Provider should be OK for the azure ad provider(?):
<img width="429" alt="image" src="https://user-images.githubusercontent.com/22298646/222677989-f4287751-8900-466e-8c4a-df558620aeaa.png">

**Reply by sventorben** [reply]
Can you please double check that you are trying to authenticate via the correct client?
I am asking, because your flow is only bound to specific clients and is not the default.

**Reply by sventorben** [reply]
> username and password as sub-flow??

Yes, because you cannot mix required and alternative flows/executions at the same level. The username/password form does cannot be configured as an alternative itself, so it needs to be wrapped in an alternative sub-flow.

**Reply by schenardi** [reply]
> Can you please double check that you are trying to authenticate via the correct client? I am asking, because your flow is only bound to specific clients and is not the default.

yes, the client is correct > checked via changes at the new flow that break the whole sign-in screen....

**Reply by schenardi** [reply]
> Yes, because you cannot mix required and alternative flows/executions at the same level. The username/password form does cannot be configured as an alternative itself, so it needs to be wrapped in an alternative sub-flow.

Can you explain a bit more about that subflow? I'm somehow missing something... On the image I cannot see the subflow itself. 

Mabey the other view helps:
<img width="1277" alt="image" src="https://user-images.githubusercontent.com/22298646/222790864-d352c3a5-0dc1-4126-bf77-1e446c5a3dd5.png">

**Reply by sventorben** [reply]
> Can you explain a bit more about that subflow? I'm somehow missing something... On the image I cannot see the subflow itself. 

The subflow is only there to be able to use the username password form as an alternative to the cookie and home idp discovery authenticators. If you check the requirement options for the username and password form, it will only show required and disabled. When wrapping it in a subflow, the subflow can be made an alternative.

Would you mind exporting your realm and posting it here?

**Reply by schenardi** [reply]
> > Can you explain a bit more about that subflow? I'm somehow missing something... On the image I cannot see the subflow itself.
> 
> The subflow is only there to be able to use the username password form as an alternative to the cookie and home idp discovery authenticators. If you check the requirement options for the username and password form, it will only show required and disabled. When wrapping it in a subflow, the subflow can be made an alternative.
> 
> Would you mind exporting your realm and posting it here?

took out a lot of info, hope you can import it. Otherwise give me please an email or contact me at markus.schenardi(at)laixo.ch
[realm-export-json.txt](https://github.com/sventorben/keycloak-home-idp-discovery/files/10898840/realm-export-json.txt)

**Reply by sventorben** [reply]
Yes, I can import it without issues.

You should see a warning in the logs that explains the problem you have:

```
WARN  [org.keycloak.authentication.DefaultAuthenticationFlow] (executor-thread-7) REQUIRED and ALTERNATIVE elements at same level! Those alternative executions will be ignored: [auth-cookie, home-idp-discovery, null]
```

You need to add the `Username Password Form` to the `username and password` sub-flow, but you have it on the same level. Click the plus icon (`+`) beside the sub-flow and click `Add step` to add the `Username Password Form` authenticator.

![image](https://user-images.githubusercontent.com/12183470/223200424-3b06f1dc-37b3-44d1-9e26-348e25074107.png)

**Reply by schenardi** [reply]
Thanks for your assistance!

Sorry, I tried that substep before but sent an older version. Behaviour changes a little bit, but I would expect to be sent to the azure provider after entering my email. And I would expect not to see the Azure Buttons. 

<img width="368" alt="image" src="https://user-images.githubusercontent.com/22298646/223337552-50ae557e-9064-4d8b-8118-8708d9b63abb.png">

<img width="385" alt="image" src="https://user-images.githubusercontent.com/22298646/223337599-461a3b19-59fc-4d6f-b18e-c228c33dbfe2.png">

**Comment by sventorben** [comment, marked as answer]
If you don't want the Azure Button, make sure you hide it in the identity provider configuration.

![image](https://user-images.githubusercontent.com/12183470/223351683-3d8f5de2-747b-4985-a224-93cdc2c93309.png)

You are mostlikey seeing the password form, because there exists a local user with that username/email and the email has not been verified. 

From the docs:
> the authenticator will only forward the user if the email has been verified

You should also see a warning in your logs like this:

```
WARN  [de.sventorben.keycloak.authentication.hidpd.DomainExtractor] (executor-thread-25) Email address of user 'f2506946-dbea-4feb-8b85-34f2546ac75b' is not verified
```

**Reply by schenardi** [reply]
Thanks a lot! 

Three parts for the solution:
* mark users (automatically) as verified once they are generated from Azure automatically (force mode)
* disable the buttons in the advanced provider settings
* define username/password form as substep of username/password (please adjust the image in the documentation with indention of the last step

<img width="871" alt="image" src="https://user-images.githubusercontent.com/22298646/223388505-bca27332-f0da-4cf6-997f-6a142d8f7ad5.png">

**Comment by ryancamp1** [comment]
Hi @sventorben 
I have a situation here where my Azure instance for the HomeIdp allows users to login with an email or upn.
![image](https://github.com/user-attachments/assets/240a979a-c984-4166-9483-b65cf9dd5a9f)
Is there a way to add multiple user attributes , or would we need to have a modified flow that caters for both instances?

**Reply by sventorben** [reply]
Hey @ryancamp1,
I think you should be good by just using email in this case and add the UPN domain/realm to the list of domains for the Azure IdP.
